### DISCUSSION TITLE
Using an npm module like react-hook-form

### DISCUSSION THREAD

**Discussion by cloudcompute** [remix-run/remix#1632] [Q&A] [answered]
Hi

Does Remix allow us to use a popular module like react-hook-form which makes validation using Zod like libraries very easy?

Also, I hope that we can use React components like Numeric Input, Credit Card Input, Multi Select etc. instead of using the basic HTML input elements?

Kindly confirm.
Thanks.

**Comment by hollandThomas** [comment, marked as answer]
# Validation

`react-hook-form` has its own implementation around forms, [useForm](https://react-hook-form.com/api/useform). That does not pair well with Remix, but the good news is: it does not need to! In your action, you've got access to a `FormData` object. You can use Zod directly to validate it (see [this issue](https://github.com/remix-run/remix/issues/539) for example):

```ts
// utils/form.ts
import type { ZodObject, ZodRawShape } from "zod";

export default function getFormData<T extends ZodRawShape>(
  form: FormData,
  inputSchema: ZodObject<T>
) {
  const keys = Object.keys(inputSchema.shape);
  let returnObject: Record<string, any> = {};
  keys.forEach((item) => {
    returnObject[item] = form.get(item);
  });

  return inputSchema.parse(returnObject);
}
```

```ts
// routes/login.tsx
import type { ActionFunction } from "remix";
import { z } from "zod";

const LoginObjectSchema = z.object({
  username: z.string(),
  password: z.string().min(5),
  redirectTo: z.string().optional(),
});

export const action: ActionFunction = async ({ request })=> {
  const formData = await request.formData();
  const { username, password, redirectTo } = getFormData(
    formData,
    LoginObjectSchema
  );
 // ...
};
```

# Custom Inputs

Many components (such as [Radix's Radiogroup](https://www.radix-ui.com/docs/primitives/components/radio-group)) can be used out-of-the-box since they ship with a visually hidden native `<input>` themselves. Notice the `<input type="radio">` and its attributes in this example. It's invisible to the end-user (both visually and for assistive technology) but is used functionally when submitting the form:

```html
<form>
  <div role="radiogroup" dir="ltr" aria-label="View density" tabindex="0" style="outline: none;">
    <div class="c-dhzjXW c-dhzjXW-iiDXXMa-css">
      <button type="button" role="radio" aria-checked="true" data-state="checked" value="default" id="r1" class="c-bVSjhL" tabindex="0" data-radix-collection-item="">
        <span data-state="checked" class="c-kYhLgI">
        </span>
      </button>
      <input type="radio" aria-hidden="true" tabindex="-1" value="default" checked="" style="transform: translateX(-100%);position: absolute;pointer-events: none;opacity: 0;margin: 0px;width: 25px;height: 25px;">
      <label for="r1" class="c-fBGwdq">Default</label>
    </div>
  </div>
</form>
```

For components where that's not the case, you can use an [\<input type="hidden">](https://www.w3schools.com/tags/att_input_type_hidden.asp) that syncs state with your custom input-component and is used to transmit form data on submit.

**Reply by cloudcompute** [reply]
Thanks. I got to know about it, I hope that in some future release of Remix, we will be able to use any React component library out-of-the-box.

**Comment by AndrewOttavianoAbb** [comment]
@cloudcompute thanks for bringing this up. We created a Form component based heavily on RHF and Yup (for validation). It would be really nice if there was a we could use the Remix Form component as a wrapper or something around our Form component. The HTML default validation is annoyingly bad and we want to take advantage of client side JS to display nice pretty validation messages 😁
